java/fury-core/src/main/java/org/apache/fury/serializer/collection/FuryArrayAsListSerializer.java

```diff
@@ -43,4 +47,91 @@ public Collection newCollection(MemoryBuffer buffer) {
     setNumElements(numElements);
     return new ArrayAsList(numElements);
   }
+
+  /**
+   * A List which wrap a Java array like `java.util.Arrays.ArrayList`, but allow to replace wrapped
+   * array. Used for serialization only, do not use it in other scenarios.
+   */
+  @Internal
+  public static class ArrayAsList<E> extends AbstractList<E>
+      implements RandomAccess, java.io.Serializable {
+    private static final Object[] EMPTY = new Object[0];
+
+    private E[] array;
+    private int size;
+
+    @SuppressWarnings("unchecked")
+    public ArrayAsList(int size) {
+      array = (E[]) new Object[size];
+    }
+
+    @Override
+    public int size() {
+      return size;
+    }
+
+    @Override
+    public boolean add(E e) {
+      array[size++] = e;
+      return true;
+    }
+
+    @Override
+    public E get(int index) {
+      return array[index];
+    }
+
+    @SuppressWarnings("unchecked")
+    public void clearArray() {
+      size = 0;
+      array = (E[]) EMPTY;
+    }
+
+    public void setArray(E[] a) {
+      array = a;
+      size = a.length;
+    }
+
+    public E[] getArray() {
+      return array;
+    }
+
+    @Override
+    public E set(int index, E element) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int indexOf(Object o) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+      throw new UnsupportedOperationException();
+    }
```

Comment:
This seems to be the effective changes beyond moving the code from a standalone file to a nested class.

But such methods are quite common and I suppose every Java developer can write:

```java
    @Override
    public boolean add(E e) {
      array[size++] = e;
      return true;
    }

    @Override
    public E get(int index) {
      return array[index];
    }
```

If such code can have compliance issues, where is the unique origin for all the code in such pattern?

Reply:
I just removed the implementation of methods which won't be invoked by Fury currently.

But to be honest, I think Java developer  will write similar code if he want to implement a List based on an Array.